jscomp/syntax/ast_utf8_string.ml

```diff
@@ -17,129 +17,230 @@
+    begin
       Buffer.add_char buf a0;
       Buffer.add_char buf a1;
       Buffer.add_char buf a2;
-      Buffer.add_char buf a3;  
-      check_and_transform loc buf s  (offset + 4) s_len 
-    end 
-  else 
+      Buffer.add_char buf a3;
+      check_and_transform loc buf s  (offset + 4) s_len
+    end
+  else
     Location.raise_errorf ~loc "%c%c%c%c is not a valid unicode point"
-      a0 a1 a2 a3 
+      a0 a1 a2 a3
 (* http://www.2ality.com/2015/01/es6-strings.html
    console.log('\uD83D\uDE80'); (* ES6*)
    console.log('\u{1F680}');
-*)   
\ No newline at end of file
+*)
+
+type interpo = Text of string | Delim of string
+
+type ('a, 'b) either = Left of 'a | Right of 'b
+
+let consume_text s start_index =
+  let rec _consume_text s index last_char new_word =
+    if index = String.length s then new_word, String.length s
+    else begin
+      match s.[index] with
+      | '\\' -> (if index + 1 = String.length s then "", index else
+                   match s.[index+1] with
+                   | '\\' -> _consume_text s (index+2) ' ' (Ext_string.append_char new_word '\\')
+                   | '$' -> _consume_text s (index+2) ' ' (Ext_string.append_char new_word '$')
+                   | c -> _consume_text s (index+1) '\\' (Ext_string.append_char new_word '\\'))
+      | '$' -> (new_word, index)
+      | c -> _consume_text s (index + 1) c (Ext_string.append_char new_word c)
+    end
+  in _consume_text s start_index ' ' ""
+
+let consume_delim s start_index =
+  let with_par = ref false in
+  let rec _consume_delim s index ident =
+    if index = String.length s then (if !with_par = true then (None, index) else (Some ident, index))
+    else
+      match s.[index] with
+      | '(' -> (if !with_par = false then (with_par := true; _consume_delim s (index+1) ident) else (None, index))
+      | ')' -> (if !with_par = false then (None, index + 1) else (with_par := false; (Some ident, index+1)))
+      | '$' -> (_consume_delim s (index+1) ident)
+      | 'a' .. 'z' | 'A' .. 'Z' | '0' .. '9'| '_' ->_consume_delim s (index+1) (Ext_string.append_char ident s.[index])
+      | _ -> if !with_par = false then (Some ident, index) else (None, index + 1)
+  in match s with
+  | "" -> (Some "", start_index)
+  | _ -> if start_index = String.length s then (Some "", start_index)
+    else (if s.[start_index] <> '$' then (None, start_index)
+          else _consume_delim s start_index "")
+
+
+let compute_new_loc (loc:Location.t) s = let length = String.length s in
+  let new_loc =
+    {loc with loc_start = {loc.loc_start with pos_cnum = loc.loc_end.pos_cnum};
+              loc_end = {loc.loc_start with pos_cnum = loc.loc_end.pos_cnum + length}}
+  in new_loc
+
+let error_reporting_loc (loc:Location.t) start_index end_index =
+  let new_loc =
+    {loc with loc_start = {loc.loc_start with pos_cnum = loc.loc_start.pos_cnum + start_index};
+              loc_end   = {loc.loc_end   with pos_cnum = loc.loc_start.pos_cnum + end_index }} in new_loc
```

Comment:
we should record offset, instead of creating new loc each time, otherwise, it would do lots of allocation